rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,218 @@
+ignored. With the same backfill mechanism, the application should now be able to
+lower the retention height below the current base height (whether to accommodate
+changing consensus parameters, app specific needs, or a local config).
+This accommodates the use case of a truncated node that wishes to become an
+archive node.
+
+#### Choosing between Strong or Weak Gaurantees about Data
+
+It can be said previously, that block pruning, with respect to the application, 
+offered strong gaurantees about data i.e. if the application pruned to height 100, 
+the operation was synchronous and thus the application knew it would prune to 100 
+before the next height. This meant that the application always knew the height and 
+base of the blockchain at every step. The opposite, weak gaurantees, implies that 
+the application can't gaurantee that what it asked from Tendermint, with respect 
+to data, has happened.
+
+The downside of strong gaurantees is that it is blocking. If a validator decided 
+to prune a million blocks at once, the validator could potentially fall behind. 
+There's an argument then that pruning (and for that matter reverse sync) shouldn't 
+lie on the consensus critical path but rather be an asynchronous background process 
+of the node.
+
+Further to this argument, nodes will not be able to prune past the evidence age 
+regardless of the retain height that the application requests. Finally, applications 
+may ask to fetch blocks back to a height that none of the nodes peers have. 
+The mechanism must ensure termination even if it doesn't reach the height requested 
+by the app.
+
+Taking all these points into consideration, ReverseSync and pruning should be
+operations that run asynchronously in the background. Application can be informed
+of the current base via `RequestEndBlock`:
+
+```proto 
+message RequestEndBlock {
+  Height uint64
+  Base uint64
+}
+```
+
+### Verification
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. This then can be followed with
+using the hashes in the trusted header to validate the other block and state
+info. This is recursively done till the backfill height.  
+
+### Design
+
+This section will provide a high level picture of the design with the specifics being
+addressed in a following ADR. The reverse mechanism requires the gossiping of two new
+messages to allow the passsing of state data:
+
+```proto
+message StateDataRequest {
+  uint64 height = 1;
+}
+```
+
+```proto
+message StateDataResponse {
+  uint64 height = 1;
+  ValidatorSet validator_set = 2;
+  ConsensusParams consensus_params = 3;
+  ABCIResponses abci_responses = 4;
```

Comment:
not sure that all applications store this info.

Reply:
I think we may have covered this in the call, but just for a written record here, I will say it here. These three data structures are what is saved in the state store. We won't be querying the application at all when we reverse sync. Executing blocks, so far, can only happen going forwards so we rely on other nodes' state stores to send us the relevant data.